### PR TITLE
ci(github): add report-only PR risk classifier

### PR DESCRIPTION
**Pull request by Audacity88**
## Summary

- **Base branch:** `master`
- **What changed and why:** Add `.github/risk-labeler.yml` as the checked-in source for automatic high-risk path and changed-line evidence used by the new report-only classifier.
- **What changed and why:** Add `pr-risk-labeler.yml`, which runs on PR updates, relevant risk/security label changes, and manual dispatch to produce a proposed `risk:*` result without changing labels, statuses, ProjectV2 fields, or approvals.
- **What changed and why:** Add `scripts/github/pr_risk_label.py` and focused tests for high-risk globs, deterministic changed-line escalation, low-risk docs/fixtures, ordinary medium-risk changes, `risk:manual`, `domain:security`, fail-closed API/source handling, summary Markdown hardening, and the #9530 test-only exception.
- **What changed and why:** Update maintainer CI, label, and PR workflow docs so the report-only phase is visible and separated from future risk-label mutation.
- **Scope boundary:** This is Phase 2 only. It does not mutate `risk:*`, enforce approval counts, create commit statuses, update ProjectV2, add required checks, revive #10461 enforcement, or execute PR code in an elevated context. #9345 remains open for the later mutation phase.
- **Blast radius:** New read-only GitHub Actions work on PR lifecycle events and manual dispatch, plus maintainer-facing summary output. The workflow reads GitHub PR metadata and trusted repository files; it does not check out, build, import, source, or execute untrusted pull-request code.
- **Linked issue(s):** Related #9345. This implements the report-only risk-classifier phase of that issue, while leaving automatic risk mutation for a later maintainer decision. Related #9530 and #9990 define the accepted risk policy this reports against. Related #9867 carries the separate deterministic size-label phase.
- **Labels:** `type:ci`, `risk:high`, `size:XL`, `ci`, `docs`, `scripts`, `distinguished contributor`, `domain:ci`

### What this does, simply

PR risk labels tell reviewers how much care a pull request needs. They are especially important for changes that touch security, release, governance, toolchain, or other trust boundaries. Today that judgment is manual, so labels can drift after a PR changes.

This PR adds a read-only report that says what risk label the current PR diff appears to need and why. It gives maintainers evidence to compare with the labels already on the PR, but it does not change those labels. That matters because a wrong automatic downgrade could make a risky PR look easier to merge than it is.

The workflow runs the classifier from trusted repository code, reads GitHub metadata about the PR, and writes a job summary. It treats uncertain evidence conservatively. High-risk configured paths stay high unless the narrow #9530 test-only proof applies. The policy also escalates deterministic changed-line signals for workflow permission expansion, secret access, OIDC, artifact publication, release behavior, elevated `pull_request_target`, toolchain install or container baselines, release-floor changes, and likely WIT contracts.

If a high-risk Rust file only changes existing `#[cfg(test)]` code, the report can recommend `risk:medium` under the #9530 rule, but only when the complete patch and base/head source prove the changed lines stayed inside the existing test-only item. Missing patches, source mismatches, conditional-compilation changes, renamed/added/removed files, mixed production changes, too many source files, structural brace changes, and unavailable content-sensitive workflow evidence stay `risk:high`.

`risk:manual` is reported as a hard freeze for future automation, and `domain:security` is reported separately from the proposed `risk:*` result. The report can therefore surface mismatches without taking over maintainer judgment.

### Scope and maintenance tradeoff

This is an XL change because the risk proof needs a real parser-shaped helper and regression suite, not because the workflow mutates much. Most of the size is Python tests and conservative handling around incomplete GitHub API evidence.

The added maintenance cost is intentional but bounded: the report has one JSON-compatible policy file, one standalone standard-library Python classifier, one workflow, and maintainer docs. I kept it separate from the existing size-label writer because this phase has a different trust boundary: it must stay read-only and auditable until maintainers approve mutation rules from report-only evidence.

The most complex piece is the #9530 exception. It exists so test-only edits inside high-risk Rust paths are not always over-classified, but the code is deliberately biased toward `risk:high` whenever the proof is incomplete or structural.

### How you can test (when useful)

- **Reviewer testing requested?** Yes, after this lands. Before merge, reviewers can run the classifier script locally with a read-only token against representative PRs; GitHub Actions manual dispatch for this new workflow becomes useful once the workflow file exists on `master`.
- **Interface(s) exercised:** GitHub Actions workflow summary and stdout JSON output.
- **Setup / preconditions:** For local script testing, set `GH_TOKEN` to a read-only GitHub token. For Actions testing, run `workflow_dispatch` after this workflow lands on `master`.
- **Steps to run:** Before merge, run `python3 scripts/github/pr_risk_label.py --repository zeroclaw-labs/zeroclaw --pr-number <number> --policy .github/risk-labeler.yml --summary /tmp/pr-risk-report.md`. After merge, open `PR Risk Report` in GitHub Actions, run manual dispatch with a target PR number, then inspect the job summary and JSON stdout.
- **Expected on this branch (after):** The job reports `report_only: true`, a proposed risk label, matching high-risk evidence when applicable, current live risk labels, `risk:manual` freeze state, `domain:security` state, #9530 exception detail, mismatches, and `mutations_attempted: false`. Untrusted filenames, labels, rule names, and error text render as inert Markdown in the human summary, and JSON appears as an indented block rather than a fence that a filename can terminate.
- **Prior behavior on `master` (before):** No `PR Risk Report` workflow exists, so maintainers have no automated report-only risk evidence to compare with live PR labels.

### How I tested

- **CI checks relied on and why they cover this change:** Required CI is starting on current head `0dec7853382ea696287f19b1033eec4cf83eab4e`; this body does not claim exact-head CI green yet. Local validation covered the classifier, policy parsing, summary Markdown hardening, workflow YAML parse, docs gates, public-comment hygiene, and diff integrity. The latest pushed repair tightens the remaining classifier edge cases: #9530 now requires real structural Rust `#[cfg(test)]` attributes, scalar/quoted/flow `pull_request_target` workflow events are recognized, deletion-only removal of explicit workflow permission restrictions escalates, and inert command text no longer counts as secret or release evidence.
- **Known CI coverage gap, if any:** `actionlint` is unavailable locally, and live GitHub Actions manual dispatch of this newly added workflow cannot run until the workflow exists on `master`. Manual dispatch remains the intended report-only comparison step before future risk mutation.
- **Commands run and tail output:**

```text
PYTHONDONTWRITEBYTECODE=1 python3 -m unittest discover -s scripts/github -p 'pr_risk_label_test.py'
Ran 31 tests in 3.049s
OK

PYTHONDONTWRITEBYTECODE=1 python3 -m unittest discover -s scripts/github -p '*_test.py'
Ran 100 tests in 8.902s
OK

python3 -m compileall -q scripts/github/pr_risk_label.py scripts/github/pr_risk_label_test.py
# no output

python3 -m json.tool .github/risk-labeler.yml >/dev/null
# no output

ruby -e 'require "yaml"; ARGV.each { |path| YAML.load_file(path); puts "parsed #{path}" }' .github/workflows/ci.yml .github/workflows/pr-risk-labeler.yml .github/workflows/windows-tests.yml .github/workflows/pr-size-labeler.yml .github/workflows/release-stable-manual.yml
parsed .github/workflows/ci.yml
parsed .github/workflows/pr-risk-labeler.yml
parsed .github/workflows/windows-tests.yml
parsed .github/workflows/pr-size-labeler.yml
parsed .github/workflows/release-stable-manual.yml

DOCS_FILES=$'.github/workflows/master-branch-flow.md\ndocs/book/src/maintainers/ci-and-actions.md\ndocs/book/src/maintainers/labels.md\ndocs/book/src/maintainers/pr-workflow.md' BASE_SHA=$(git merge-base origin/master HEAD) bash scripts/ci/docs_quality_gate.sh
Summary: 0 error(s)

DOCS_FILES=$'.github/workflows/master-branch-flow.md\ndocs/book/src/maintainers/ci-and-actions.md\ndocs/book/src/maintainers/labels.md\ndocs/book/src/maintainers/pr-workflow.md' BASE_SHA=$(git merge-base origin/master HEAD) bash scripts/ci/docs_links_gate.sh
Ran 6 tests in 3.294s
OK
Collected 0 added link(s) from 4 docs file(s).
No added links detected in changed docs lines.

bash scripts/ci/comment_hygiene_gate.sh
Comment hygiene gate passed.

git diff --check origin/master...HEAD
# no output

GH_TOKEN=<read-only-token> python3 scripts/github/pr_risk_label.py --pr-number 10861 --policy .github/risk-labeler.yml --repository zeroclaw-labs/zeroclaw
Proposed risk: risk:high
High-risk evidence: \.github/risk\-labeler\.yml (path: \.github/risk\-labeler\.yml), \.github/workflows/pr\-risk\-labeler\.yml (path: \.github/workflows/pr\-risk\-labeler\.yml), \.github/workflows/pr\-risk\-labeler\.yml (content: elevated pull\_request\_target; lines: 4), scripts/github/pr\_risk\_label\.py (path: scripts/github/pr\_risk\_label\.py)
GitHub label mutation: none
Commit status mutation: none
Approval-count enforcement: none
```

- **Beyond CI, what did you manually verify?** I inspected the workflow trust boundary: `pull_request_target` uses only read permissions, fetches the classifier and policy from the trusted workflow revision, and does not check out or run PR code. I also checked open PR overlap: the current open docs/CI PRs touch neighboring maintainer docs or CI jobs, but none implements `pr-risk-labeler.yml`, `.github/risk-labeler.yml`, or `scripts/github/pr_risk_label.py`. After the refresh, I verified that `origin/master...HEAD` still contains the intended 9-file PR diff. I also ran the classifier against this PR through the GitHub API at head `0dec7853382ea696287f19b1033eec4cf83eab4e`; it returned `risk:high`, preserved `mutations_attempted: false`, reported path evidence for `.github/risk-labeler.yml`, `.github/workflows/pr-risk-labeler.yml`, and `scripts/github/pr_risk_label.py`, and reported changed-line evidence for the real elevated `pull_request_target` workflow event without the earlier docs/test-fixture false positives.
- **Visual interface evidence:** N/A; this adds a GitHub Actions summary and JSON report, not a visual layout or interactive UI.
- **Tested revision, interface, and evidence path:** `0dec7853382ea696287f19b1033eec4cf83eab4e`; local command output above plus live classifier smoke against PR #10861. Future GitHub Actions summary evidence becomes available after the workflow lands on `master`.
- **Screenshot or exact-output evidence:** Exact local command output above.
- **Action or changed state and observed result:** The classifier proposes risk labels and emits report-only evidence in tests and live API smoke; no GitHub label/status/project mutation occurs.
- **If any command was intentionally skipped, why:** `actionlint` was skipped because it is not installed locally. Live manual-dispatch workflow execution is not possible before this workflow lands on `master`.

## Security & Privacy Impact (required)

- New permissions, capabilities, or file system access scope? Yes. The new workflow declares `contents: read` and `pull-requests: read` only. It uses `pull_request_target`, so the mitigation is that it fetches trusted repository code and policy, reads PR metadata, and never checks out or executes untrusted PR code.
- New external network calls? Yes. The workflow calls the GitHub API to read PR metadata, changed files, and trusted source text for the bounded #9530 proof. It does not call third-party services.
- Secrets / tokens / credentials handling changed? No. The workflow uses the ephemeral `GITHUB_TOKEN` with read-only permissions and does not add, store, print, or transform secrets.
- PII, real identities, or personal data in diff, tests, fixtures, or docs? No. Fixtures use synthetic paths, labels, and SHA values.
- Prompt injection or untrusted model-visible text introduced/changed? No. No model prompt is introduced. Untrusted PR filenames, labels, rule names, and errors are escaped before Markdown summary output; the JSON report uses an indented block rather than a Markdown fence; PR code is never executed.
- If any `Yes`, describe the risk and mitigation: The risk is giving an elevated-event workflow too much power over fork PRs. This PR limits the workflow to read-only permissions, trusted-file fetches, GET-only API use, report-only output, inert Markdown summary rendering, and fail-closed handling for malformed or incomplete evidence.

## Compatibility (required)

- Backward compatible? Yes.
- Config / env / CLI surface changed? No.
- Rust/MSRV/toolchain floor changed? No.
- If backward compatibility is `No` or either surface/floor question is `Yes`: N/A.

## Rollback (required for medium/high-risk PRs)

- **Fast rollback command/path:** After merge, revert the landed squash commit from `master` with `git revert <landed-squash-sha>` and push through the normal maintainer path. If the workflow must be disabled before the revert lands, make an emergency follow-up commit that removes `.github/workflows/pr-risk-labeler.yml` or restricts it to `workflow_dispatch` only, then follow with the full revert.
- **Feature flags or config toggles:** None. This is report-only; removing or disabling the workflow is the rollback path.
- **Observable failure symptoms:** `PR Risk Report` jobs fail, job summaries report incorrect proposed risks, `mutations_attempted` is anything other than `false`, or summaries/API output show active Markdown or malformed evidence instead of failing closed.

## Supersede Attribution (required only when `Supersedes #` is used)

- Superseded PRs + authors (`#<pr> by @<author>`, one per line): N/A.
- Scope materially carried forward: N/A.
- `Co-authored-by` trailers added in commit messages for incorporated contributors? No.
- If `No`, why (inspiration-only, no direct code/design carry-over): This does not use `Supersedes #`; no contributor PR code is materially carried forward.

### CODE REVIEW

**Comment by IftekharUddin** [comment]
Two blockers for this report-only classifier:

1. **Untrusted Markdown in the `pull_request_target` summary.** `matching_evidence.path` comes from PR-controlled filenames, and `human_summary()` interpolates it into Markdown after `summary_text()` only HTML-escapes and escapes backticks/newlines. A filename under a high-risk glob can still contain Markdown link/image syntax and render active content in the job summary. Please render paths as inert text with complete Markdown-safe encoding (or a safely delimited code representation) and add a regression test using link/image syntax.

2. **Missing accepted escalation rules.** #9345 Phase 2 requires explicit workflow/toolchain escalation, and #9990 says the report must classify up when deterministic diff evidence shows workflow permission, secret, OIDC, artifact-publication, release-behavior, or elevated `pull_request_target` changes, and when the toolchain/install/container baseline or release floor changes. The classifier currently defaults non-high-path/non-low-path changes to `risk:medium` without inspecting those diff contents; for example, a permission expansion in an otherwise unlisted workflow or a changed container toolchain baseline can be reported as medium. Please add deterministic rules and positive/negative fixtures, or make a maintainer decision to revise the accepted policy and clarify this slice’s scope before treating it as the report-only phase.

**Comment by Audacity88** [comment]
@IftekharUddin I pushed `cef4da0855c3d649c27022b4dc2b6fbd5759de26` to address the two blockers.

The summary output now makes PR-controlled text inert before writing the job summary. Filenames, labels, rule names, and error text are Markdown-escaped, and the JSON report is written as an indented block instead of a fenced block, so link/image syntax and triple-backtick filenames cannot render active Markdown or terminate the report structure.

The risk policy now also has deterministic changed-line escalation rules for workflow permission expansion, secret access, OIDC, artifact publication, release behavior, elevated `pull_request_target`, toolchain install or container baselines, release-floor changes, and likely WIT contracts. The classifier reports that content evidence alongside path evidence and stays `risk:high` when a content-sensitive patch is unavailable or malformed.

I added regressions for the Markdown cases, all nine changed-line escalation rules, negative read-only/docs cases, and missing content-sensitive patches. I also refreshed the PR body with the current tested head, updated validation counts, live classifier smoke, the pending exact-head CI status, and an executable rollback path.

Local validation passed: focused risk tests (`29`), all GitHub helper tests (`93`), Python compileall, policy JSON parse, workflow YAML parse, docs quality, docs links, comment hygiene, diff checks, and live classifier smoke against this PR. Exact-head GitHub CI is running on `cef4da0855c3d649c27022b4dc2b6fbd5759de26`; no failures were present in the first readback, and I am not claiming the required gate is green yet.

**Comment by Audacity88** [comment]
Pushed `7a93fae8a291852570dd39b503c7f8c21e4a968a` to address the current-head review findings.

The branch is refreshed onto current `master` with a normal merge commit. GitHub now reports the PR as mergeable, and the PR diff against `origin/master` is still the intended 9 files: the risk policy, risk workflow, CI/docs updates, classifier, and focused tests. Fresh required CI is running on the new head, so I am not claiming exact-head CI green yet.

The changed-line classifier now treats workflow-only semantic rules as workflow YAML evidence, not as prose evidence from neighboring docs. It no longer escalates `.github/workflows/master-branch-flow.md` release-table text, and it skips test/fixture paths such as `scripts/github/pr_risk_label_test.py` for content-rule escalation. I also narrowed workflow permission matching to known GitHub permission keys so `name: write` does not look like a permission expansion.

The workflow YAML cases from the review are now covered: `permissions: {contents: write}`, `permissions: {id-token: write}`, and `on: [pull_request_target]` all escalate. The existing block-style workflow permission, OIDC, and `pull_request_target` cases still escalate. Summary escaping now also escapes `~`, so `~~untrusted~~` renders as inert text rather than strikethrough.

Local validation passed on the pushed head: focused risk tests (`29`), all GitHub helper tests (`98`), Python compileall, policy JSON parse, Ruby workflow YAML parse for `ci.yml`, `pr-risk-labeler.yml`, and `windows-tests.yml`, docs quality, docs links, comment hygiene, diff checks, and live classifier smoke against this PR. The live classifier smoke returns `risk:high` with `mutations_attempted: false` and now reports only the real path evidence plus the real elevated `pull_request_target` workflow event, without the prior docs/test-fixture false positives.

I also refreshed the PR body with the current tested head, mergeability state, pending exact-head CI language, updated validation counts, and current live classifier smoke.

**Comment by Audacity88** [comment]
I pushed `46fc0ab525e5edaf4f24a3ca25309d5b2a72e1af` for the latest review findings.

Content-sensitive evidence now fails closed when the parsed patch line counts do not match GitHub's file metadata. That covers the case where a patch is syntactically valid but incomplete: the classifier reports `content-sensitive diff unavailable` and keeps the proposed risk at `risk:high` instead of trusting partial content evidence.

Workflow YAML evidence is also more context-aware. The classifier strips YAML comments outside quoted strings, ignores metadata-only `name` and `run-name` lines, and recognizes quoted flow-map permissions such as `permissions: {"contents": "write"}` and `permissions: {'id-token': 'write'}`. That preserves the real workflow permission and OIDC escalations while avoiding comment/metadata false positives like `# secrets: none` and `name: release docs`.

I added focused regressions for the truncated-patch fail-closed case, quoted workflow permission forms, metadata-only workflow names, and comment-only secret mentions. Local validation passed on this head: focused risk tests (`30`), all GitHub helper tests (`99`), Python compileall, policy JSON parse, workflow YAML parse, docs quality, docs links, comment hygiene, diff checks, and live classifier smoke against this PR.

The PR body draft has also been refreshed locally for head `46fc0ab525e5edaf4f24a3ca25309d5b2a72e1af`, the current label snapshot including `distinguished contributor`, and the updated validation counts. Exact-head GitHub CI has started; I am not claiming it green yet.

File: scripts/github/pr_risk_label.py [hunk 1]
```diff
@@ -0,0 +1,1073 @@
...
+        index += 1
+    return "".join(output)
+
+
+def cfg_test_ranges(source: str) -> list[tuple[int, int]]:
+    lines = source.splitlines()
+    structural_lines = rust_structural_text(source).splitlines()
+    ranges: list[tuple[int, int]] = []
+    for index, line in enumerate(lines):
+        if not CFG_TEST_RE.match(line):
```
Comment (JordanTheJet) [inline scripts/github/pr_risk_label.py:618]: This matches original text, so a standalone #[cfg(test)] inside a block comment can create a test-only range over the following production return statement. Match structural source and require a real existing test item; include commented/multiline-literal attributes as negative fixtures.

File: scripts/github/pr_risk_label.py [hunk 2]
```diff
@@ -0,0 +1,1073 @@
...
+        }
+    if not parsed_patch_counts_match(item, old_changed, new_changed):
+        return {
+            "path": path,
+            "content_rules": ["content-sensitive diff unavailable"],
+        }
+
+    matches: list[str] = []
+    line_numbers: list[int] = []
+    for line_number, line in new_changed:
```
Comment (JordanTheJet) [inline scripts/github/pr_risk_label.py:790]: The new count reconciliation fixes partial evidence, but consequence detection still drops deletions. Removing a job-level read-only permissions restriction can inherit unchanged workflow-level write authority without any added line. Cover that before/after permission change; also include scalar/quoted/flow-map pull_request_target forms.

File: scripts/github/pr_risk_label.py [hunk 3]
```diff
@@ -0,0 +1,1073 @@
...
+        semantic_line = workflow_semantic_line(line)
+        if semantic_line is None:
+            return False
+        if rule.name == "workflow permission expansion":
+            return workflow_permission_write(semantic_line)
+        if rule.name == "OIDC token access":
+            return workflow_oidc_write(semantic_line)
+        if rule.name == "elevated pull_request_target":
+            return workflow_pull_request_target(semantic_line)
+        return rule.pattern.search(semantic_line) is not None
```
Comment (JordanTheJet) [inline scripts/github/pr_risk_label.py:371]: Raw rule matching still labels run: echo release as release behavior and run: echo secrets.RELEASE_TOKEN as secret access. Neither establishes that consequence. Add inert command negatives and constrain matches to actual workflow expressions/publication semantics.

**Review by JordanTheJet — CHANGES_REQUESTED** [review]
@Audacity88, I reviewed 46fc0ab525e5edaf4f24a3ca25309d5b2a72e1af against the report-only scope and accepted #9530/#9990 policy. The trusted-code and read-only API boundary remains sound. I am requesting changes for the classification defects below; the partial-patch completeness check from my earlier draft is now fixed.

### 🔴 Blocking — Commented attributes can still downgrade production changes

`cfg_test_ranges()` matches `CFG_TEST_RE` against the original source lines, rather than the comment/literal-stripped source. In a configured high-risk Rust file, this unchanged prefix is enough to produce a false test-only range:

```rust
/*
#[cfg(test)]
*/
pub fn allowed() -> bool {
    return false;
}
```

Changing only `return false;` to `return true;` is a production change. The commented attribute matches; the blanked comment terminator selects the semicolon-scanning branch; and the range extends through that return statement. Both revisions pass the range check, while the changed line has neither a brace nor a cfg attribute. The classifier can therefore apply #9530 and propose medium for a production edit.

Recognize attributes from structural source and require an actual existing test-only item before downgrading. Add apparent attributes inside block comments and multiline literals as negative cases. The current lexer regression checks braces around a genuine test item, so it does not cover this case. This is a static trace, not an executed reproduction.

### 🔴 Blocking — Workflow escalation still misses ordinary event and permission changes

The quoted flow-map permission and sequence-event repairs are present. However, `workflow_pull_request_target()` still misses scalar forms such as `on: pull_request_target` and `on: "pull_request_target"`, as well as a flow mapping. In an otherwise unlisted workflow, those changes can fall through to medium. GitHub supports a scalar single-event declaration. [Workflow syntax](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#on)

Separately, `content_evidence_for_file()` validates both patch sides but inspects only additions for consequences. If workflow-level `contents: write` is unchanged, deleting a job's explicit read-only `permissions` restriction expands its inherited authority without adding a line for these rules to match. The count-completeness repair does not fix that case.

Evaluate the relevant before/after workflow structure, or keep ambiguous consequential edits high. Add scalar/quoted/flow-map event fixtures and a deletion-only job-permission restriction fixture, alongside ordinary read-only negatives. These are deterministic changes within the accepted escalation contract.

### 🔴 Blocking — Literal command text is not evidence of secret or release access

The current comment/name filters fix the reported examples, but other workflow rules still search arbitrary command text. `run: echo release` matches “release behavior”; `run: echo secrets.RELEASE_TOKEN` matches “secret access” even though no GitHub expression accesses a secret. This confirms the latest review comment's remaining false-positive class by inspection.

Narrow the evidence to syntax that establishes the named consequence, and add inert command/string negatives alongside real secret-expression and publication positives. Report-only calibration is useful, but the report should not present a literal vocabulary match as proof of authority or secret access.

### ✅ Resolved — Completeness, escaping, and earlier workflow cases

Parsed addition/deletion counts now reconcile with GitHub metadata before content classification; a complete-looking partial patch stays high. Summary escaping covers the demonstrated link, image, fence, and strikethrough cases, and JSON is indented. Quoted permission/OIDC mappings, workflow metadata names, comments, and neighboring docs/fixtures have new coverage. These resolve the concrete earlier cases without settling the broader workflow findings above.

The [required Security job](https://github.com/zeroclaw-labs/zeroclaw/actions/runs/35562190135/job/106217034360) validates the new workflow YAML and reports `Ran 99 tests`, `OK`, covering all 30 classifier/workflow tests in this head. [CI Required Gate](https://github.com/zeroclaw-labs/zeroclaw/actions/runs/35562190135/job/106219677286) and Docs Style pass. I used those hosted results; no PR code, local tests, or Cargo ran during this review. There are no formal human request-changes reviews to supersede; @IftekharUddin's top-level escalation concern remains partially open. Old automated transport alerts concern replaced implementations.

### 🟡 Warning — Finish the evidence and interface declarations

The body now includes actual CLI output, which addresses the absence of any report excerpt. Include its JSON line or a linked complete output artifact to expose the captured SHAs, override/security state, and mutation flag rather than relying only on narrative for those fields. The unavailable pre-merge Actions dispatch remains an explicit rollout gap. Replace the stale “CI is starting” sentence with the current gate, and qualify the compatibility answer to acknowledge the new maintainer CLI/dispatch surface. The runtime surface remains unchanged.

GitHub reports this head mergeable, 15 commits behind `master`, with required CI passing. Base drift is a freshness warning, not the reason for this verdict. The label snapshot and landed-squash rollback now match the change.

### 🟢 What looks good — The report leaves the gate and human judgment intact

Permissions are content/PR reads; API calls are GET-only; executable files come from the trusted repository revision; contributor source is inert data. There is no label, status, ProjectV2, approval-count, or required-check writer. The `ci.yml` edit adds YAML validation without changing the aggregate gate's dependencies or result logic. `risk:manual` and `domain:security` remain distinct report fields. The XL classifier/proof and fixture suite form a coherent report-only slice, and `Related #9345` correctly leaves mutation for a later decision.

**Review by IftekharUddin — APPROVED** [review]
I reviewed `46fc0ab525e5edaf4f24a3ca25309d5b2a72e1af` and approve this initial report-only PR risk classifier.

### 🟢 What looks good — Strict read-only boundaries and safe trusted execution

- **Uncompromising security boundaries:** The workflow executes under minimal read-only permissions (`contents: read`, `pull-requests: read`), makes only GET requests, and never mutates repository state (no labeling, commenting, project status, or branch protection updates).
- **Execution safety under `pull_request_target`:** The classifier script (`scripts/github/pr_risk_label.py`) and policy are fetched exclusively from the trusted target revision (`$TRUSTED_REF`), never checking out or executing untrusted contributor code.
- **Report-only sandboxing:** Outputs are piped strictly to `$GITHUB_STEP_SUMMARY`, keeping human review and branch protection fully authoritative.
- **CI Required Gate:** All 42 required and auxiliary CI check suites pass on head `46fc0ab525e5edaf4f24a3ca25309d5b2a72e1af`.

### ✅ Resolved — Markdown sanitization & patch count reconciliation

Previous blockers regarding path interpolation in job summaries and patch truncation have been resolved: paths and summary fields are properly sanitized against markdown/HTML injection, and patch counts reconcile with GitHub metadata before evaluation.

### 🟡 Warning — Edge-case classifier calibration for follow-up tuning

In light of @JordanTheJet's review, the following classification edge cases are noted as non-blocking calibration items for this report-only phase:
1. **Commented test attributes:** Refine `cfg_test_ranges()` to strip block comments and string literals so commented `/* #[cfg(test)] */` attributes do not falsely delineate test ranges in high-risk files.
2. **Workflow triggers & permission removals:** Expand trigger detection to cover scalar event strings (`on: pull_request_target`), and inspect structural permission removals that expand effective privileges from top-level definitions.
3. **Command text false positives:** Constrain secret and release regex matches to GitHub Actions expressions (`${{ secrets.* }}`) rather than literal unparsed shell strings (`echo release`).

Because this initial phase is report-only and cannot mutate labels or block PRs, these heuristic edge cases are appropriate to refine in follow-up iterations as live report feedback is gathered.

**Comment by Audacity88** [comment]
@JordanTheJet I pushed `0dec7853382ea696287f19b1033eec4cf83eab4e` for the latest request-changes review.

The #9530 proof now recognizes `#[cfg(test)]` only from structural Rust source, after comments and literals are blanked. That prevents commented attributes and raw-string text from creating fake test-only ranges, and I added negative regressions for both cases.

Workflow event and permission evidence is tighter now. Scalar, quoted scalar, sequence, and flow-map `pull_request_target` event forms all escalate. The classifier also treats deletion-only removal of explicit read/none workflow permission restrictions as permission expansion, while preserving the ordinary read-only replacement negative case.

For the command-text finding, workflow secret evidence now requires a GitHub Actions secret expression or `secrets:` mapping syntax, and release behavior is limited to release-driving syntax such as `gh release create/upload`, tag filters, or version tag entries. Inert shell text like `run: echo release secrets.RELEASE_TOKEN` no longer counts as release or secret evidence.

Local validation passed on this head: focused risk tests (`31`), all GitHub helper tests (`100`), Python compileall, policy JSON parse, workflow YAML parse, docs quality, docs links, comment hygiene, diff checks, and live classifier smoke against this PR. The PR body draft has been refreshed locally for head `0dec7853382ea696287f19b1033eec4cf83eab4e` and the updated validation counts. Exact-head GitHub CI has started; I am not claiming it green yet.
